tests/test_threading.py

```diff
@@ -107,6 +107,43 @@ def test_pickle(self):
         self.assertAlmostEqual(blocked - begin, 1, delta=0.1)
 
 
+class TestMultipleContexts(TestCore):
+
+    class TestContext(object):
+        def __init__(self, event_list):
+            self._event_list = event_list
+
+        def __enter__(self):
+            self._event_list.append((self, "enter"))
+
+        def __exit__(self, type, value, traceback):
+            self._event_list.append((self, "exit"))
+
+    def setUp(self):
+        self.event_list = []
+
+        self.c1 = self.TestContext(event_list=self.event_list)
+        self.c2 = self.TestContext(event_list=self.event_list)
+
+        self.stuff = Stuff(machine_cls=MachineFactory.get_predefined(locked=True), extra_kwargs={
+            'context': [self.c1, self.c2]
+        })
+        del self.event_list[:]
+
+        self.stuff.machine.add_transition('forward', 'A', 'B')
+
+    def tearDown(self):
+        pass
+
+    def test_ordering(self):
+        self.stuff.forward()
+        # There are a lot of internal enter/exits, but the key is that the outermost are in the expected order
```

Comment:
Yes and no. Yes: Locks have to be reentrant since every function which can be called will require the look and this procedure is most likely nested. No: Afaik there is no monkey patching going on, only inheritance and therefore method overloading/overriding. What could be done is to check wether the lock has been acquired already and if that is the case skip the `LockedMethod` wrapping and return the base class method directly. This needs to be done for `LockedEvent` as well. I will have a look into it.